### DISCUSSION TITLE
How to implement a Crossplane Provider for Asynchronous API?

### DISCUSSION THREAD

**Discussion by BigGold1310** [crossplane/crossplane#5990] [Q&A]
# Implementing a Crossplane Provider for Asynchronous API

## Request for Guidance

The current implementation doesn't fit well with the "native" provider model. I'm seeking advice, tips, and tricks on how to effectively implement a provider that interacts with an asynchronous API.

Key questions:
1. How can I better handle the asynchronous nature of the workflow?
2. What's the best way to manage state during the potentially long execution times?
3. Are there alternative approaches or patterns that would be more suitable for this use case?

## Background

I'm attempting to build a Crossplane provider that interacts with an asynchronous API. This provider will manage on-premises VMs through our company's Workflow engine.

## Use Case

We want to allow the provisioning of on-premises VMs through Crossplane as managed objects. Our Workflow engine is responsible for various on-premises tasks, including VM provisioning.

## Workflow Engine Details

- The engine can execute different workflows (e.g., Create, Delete) to manage VMs.
- When a workflow is triggered, it returns a WorkflowID (tracking ID).
- This WorkflowID is not directly related to the final object (VM).
- We need to track the workflow execution until it reaches a completed state.
- Once completed, we can query for the external-name of the created object (VM name and ID).
- This logic applies to deletion and updating of VMs as well.
- Workflow execution times can vary, potentially up to 2 hours.
```mermaid
sequenceDiagram
    participant C as Crossplane
    participant P as Provider
    participant W as Workflow Engine
    participant VM as VM Infrastructure

    C->>P: Request VM Creation
    P->>W: Trigger Create Workflow
    W-->>P: Return WorkflowID
    P-->>C: Acknowledge (In Progress)

    loop Check Workflow Status
        P->>W: Query Status (WorkflowID)
        W-->>P: Return Current Status
    end

    W->>VM: Create VM
    VM-->>W: VM Created

    P->>W: Query Status (WorkflowID)
    W-->>P: Completed, Return VM Name & ID
    P-->>C: Resource Created (VM Name & ID)

    Note over C,VM: Deletion/Update process follows a similar pattern
```

## Current Implementation Attempt

I've tried implementing a "native" Crossplane provider, but faced several challenges:

1. The provider doesn't seem well-suited for tracking asynchronous execution.
2. The external-name is only set during the Create method.
4. The Observe method only persists the status object when ResourceUpToDate = true or when an error occurs in the Observe function.
5. This led to implementing workarounds, such as first claiming the resource is UpToDate, then reconciling after claiming it's not created

## Why not just migrate the whole workflow to crossplane?

This isn't that easy. The most difficult part is that we are talking here about 100s of workflows which we would need to migrate.
Our current strategy is to first onboard the resources as managed resource to crossplane. Over time we than migrate as many workflows as possible to crossplane.

**Comment by cychiang** [comment]
Hi, 

I'm not sure about how you implement your solution, but based on your description it seems to me the crossplane provider should be able to manage a workflow asynchronously. 

As you know, a Crossplane Provider is a provider for managed resource, a resource which indicate and store a state from an external resource.  

Here are my understandings from your usecase: 
- You have a workflow to manage a vm
- You will get a workflow id from the workflow when the time you send out a request from a Provider
- You will need to use this workflow id to query the progress of creation and also everything 

It sounds like a typical Crossplane Provider implementation to me: 
- A provider make a request to create something to a Cloud service provider
- A provider usually get an identifier to query the status of the request
- When it's done, write status back to managed resource in the cluster.

From your diagram, you can remove the crossplane part. Because the crossplane provider can work individually, it only care about the "managed resource" offered by a particular crossplane provider. So here is roughly idea around your question within the reconcile loop: 

- Find a workflow id from a MR -> If no workflow id means it's a new resource, then send a request to create a new resource.
  - Store workflow id in a place you want within the MR, usually will store in `status.atProvider.`
- Validate the status of a MR -> Use stored workflow id from the MR to query status and write back to MR `status.atProvider`

You don't need to finish the whole step within 1 reconcile. A reconcile loop is a place to host your business logic on how to reach to a state you want, and it's not like a for loop. The reconcile loop means the logic within the reconcile loop will be examined at any given time. [Here](https://book.kubebuilder.io/getting-started.html?highlight=reconcile%20loop#reconciliation-process) you can find a detail about the controller. The Crossplane Provider is based on top of the `controller-runtime` so it's a foundation of the Crossplane Provider.

**Reply by BigGold1310** [reply]
Hi cychiang,

Thank you for taking the time to look at my discussion. I appreciate your insights.

I agree with your approach and perspective on the provider. I've successfully implemented the full lifecycle within a standard Kubernetes controller using controller-runtime. Before delving deeper into the details, I'm curious if you know of any asynchronous implementations I could use as a reference?

**The Technical Challenge**

I'm encountering difficulties with the Crossplane reconcile implementation. Let me provide an example to clarify it a bit more:

1. **Create Function Limitation**: The interface declaration of the `Create` function and its usage within the crossplane `Reconcile` implementation only allow updating annotations. (See: [Interface](https://github.com/crossplane/crossplane-runtime/blob/main/pkg/reconciler/managed/reconciler.go#L328-L333) and [Implementation](https://github.com/crossplane/crossplane-runtime/blob/main/pkg/reconciler/managed/reconciler.go#L1175-L1184)). My understanding is that this is intended to set the `external-name` to the identifier of the created resource, which I don't know yet as I need to wait for the workflow completion.

2. **Workflow ID as External Name**: I considered writing the workflow ID to the `external-name`. However, this approach poses challenges during an Update workflow, as it doesn't care about the old workflow execution ID (temporary) but rather the created object ID.

3. **Observe Method Constraints**: Given these limitations, I attempted to update and write the resource and resource status properly within the `Observe` method. The challenge here is that the `Observe` method generally doesn't update the Resource, except when `ResourceLateInitialized` is set (see [here](https://github.com/crossplane/crossplane-runtime/blob/main/pkg/reconciler/managed/reconciler.go#L1219-L1233)). It only updates the Status of the resource and this only happens if the resource is reported to be existing as if not, I'm running into [GracePeriod](https://github.com/crossplane/crossplane-runtime/blob/main/pkg/reconciler/managed/reconciler.go#L1003-L1007) which does not update the resource at all.

Could you provide any guidance, tips, or point me towards a reference implementation for handling asynchronous operations? 

Any insights or examples of similar asynchronous implementations would be greatly appreciated. I'm particularly interested in how other providers manage long-running operations or workflows where the result isn't immediately available.
